File: Ahorn/entities/deadlyDashSwitch.jl
module FlushelineDeadlyDashSwitch

using ..Ahorn, Maple

@mapdef Entity "vitellary/deadlydashswitch" DeadSwitch(x::Integer, y::Integer, direction::String="Left", persistent::Bool=false)

const placements = Ahorn.PlacementDict(
	"Deadly Dash Switch (Left) (Crystalline)" => Ahorn.EntityPlacement(
        DeadSwitch,
        "point",
        Dict{String, Any}(
            "direction" => "Left"
        )
    ),
	"Deadly Dash Switch (Right) (Crystalline)" => Ahorn.EntityPlacement(
        DeadSwitch,
        "point",
        Dict{String, Any}(
            "direction" => "Right"
        )
    ),
	"Deadly Dash Switch (Up) (Crystalline)" => Ahorn.EntityPlacement(
        DeadSwitch,
        "point",
        Dict{String, Any}(
            "direction" => "Up"
        )
    ),
	"Deadly Dash Switch (Down) (Crystalline)" => Ahorn.EntityPlacement(
        DeadSwitch,
        "point",
        Dict{String, Any}(
            "direction" => "Down"
        )
    )
)

const directions = ["Left", "Right", "Up", "Down"]

Ahorn.editingOptions(entity::DeadSwitch) = Dict{String, Any}(
    "direction" => directions
)

function Ahorn.selection(entity::DeadSwitch)
    x, y = Ahorn.position(entity)
    dir = get(entity.data, "direction", "Left")

    if dir == "Left"
		return Ahorn.Rectangle(x - 2, y, 10, 16)
    elseif dir == "Right"
        return Ahorn.Rectangle(x, y, 10, 16)
	elseif dir == "Up"
		return Ahorn.Rectangle(x, y - 4, 16, 12)
	elseif dir == "Down"
		return Ahorn.Rectangle(x, y, 16, 12)
    end
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::DeadSwitch, room::Maple.Room)
    dir = get(entity.data, "direction", "Left")
	
	if dir == "Left"
        Ahorn.drawSprite(ctx, "objects/deadlyDashButton/dashButton00.png", 7, 8)
    elseif dir == "Right"
		Ahorn.drawSprite(ctx, "objects/deadlyDashButton/dashButton00.png", 49, 56, rot=pi)
	elseif dir == "Up"
		Ahorn.drawSprite(ctx, "objects/deadlyDashButton/dashButton00.png", 56, 7, rot=pi / 2)
	elseif dir == "Down"
		Ahorn.drawSprite(ctx, "objects/deadlyDashButton/dashButton00.png", 8, 49, rot=-pi / 2)
    end
end

end

File: Loenn/entities/deadly_dash_switch.lua
local utils = require "utils"
local enums = require "consts.celeste_enums"
local drawableSprite = require "structs.drawable_sprite"

local deadlySwitch = {}

deadlySwitch.name = "vitellary/deadlydashswitch"
deadlySwitch.placements = {}

for _,dir in ipairs(enums.dash_switch_sides) do
    table.insert(deadlySwitch.placements, {
        name = "deadlyswitch_"..string.lower(dir),
        data = {
            direction = dir,
            persistent = false,
        }
    })
end
deadlySwitch.fieldInformation = {
    direction = {
        editable = false,
        options = enums.dash_switch_sides,
    },
}

deadlySwitch.justification = {0.5, 0.5}
function deadlySwitch.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/deadlyDashButton/dashButton00", entity)
    if entity.direction == "Right" then
        sprite.rotation = math.pi
        sprite:addPosition(0, 8)
    elseif entity.direction == "Down" then
        sprite.rotation = -math.pi/2
        sprite:addPosition(8, 0)
    elseif entity.direction == "Up" then
        sprite.rotation = math.pi/2
        sprite:addPosition(8, 8)
    else
        sprite.rotation = 0
        sprite:addPosition(8, 8)
    end
    return sprite
end

function deadlySwitch.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    if entity.direction == "Right" then
        return utils.rectangle(x,y,8,16)
    elseif entity.direction == "Down" then
        return utils.rectangle(x,y,16,8)
    elseif entity.direction == "Up" then
        return utils.rectangle(x,y,16,8)
    else
        return utils.rectangle(x,y,8,16)
    end
end

return deadlySwitch